File: packages/google_mlkit_genai_speech_recognition/android/src/main/java/com/google_mlkit_genai_speech_recognition/SpeechRecognizer.java
package com.google_mlkit_genai_speech_recognition;

import android.content.Context;

import androidx.annotation.NonNull;


import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.FutureCallback;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class SpeechRecognizer implements MethodChannel.MethodCallHandler {
    private static final String CHECK_STATUS = "genai#checkStatus";
    private static final String START_RECOGNITION = "genai#startRecognition";
    private static final String STOP_RECOGNITION = "genai#stopRecognition";
    private static final String CLOSE = "genai#closeSpeechRecognizer";

    private final Context context;
    private final Map<String, Object> instances = new HashMap<>();
    private final Executor executor = Executors.newSingleThreadExecutor();

    public SpeechRecognizer(Context context) {
        this.context = context;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        String method = call.method;
        switch (method) {
            case CHECK_STATUS:
                checkStatus(call, result);
                break;
            case START_RECOGNITION:
                startRecognition(call, result);
                break;
            case STOP_RECOGNITION:
                stopRecognition(call, result);
                break;
            case CLOSE:
                closeSpeechRecognizer(call);
                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private Object initialize(MethodCall call) {
        // Speech Recognition API initialization - structure may vary
        // This is a placeholder implementation using reflection
        try {
            Object optionsBuilder = Class.forName("com.google.mlkit.genai.speechrecognition.SpeechRecognizerOptions")
                .getMethod("builder", Context.class).invoke(null, context);
            Object options = optionsBuilder.getClass().getMethod("build").invoke(optionsBuilder);
            Object speechRecognizer = Class.forName("com.google.mlkit.genai.speechrecognition.SpeechRecognition")
                .getMethod("getClient", options.getClass()).invoke(null, options);
            return speechRecognizer;
        } catch (Exception e) {
            // If reflection fails, return a placeholder
            return new Object();
        }
    }

    private void checkStatus(MethodCall call, MethodChannel.Result result) {
        String id = call.argument("id");
        Object speechRecognizer = instances.get(id);
        if (speechRecognizer == null) {
            speechRecognizer = initialize(call);
            instances.put(id, speechRecognizer);
        }

        try {
            // Speech Recognition uses checkStatus() instead of checkFeatureStatus()
            ListenableFuture<Integer> future = (ListenableFuture<Integer>) speechRecognizer.getClass()
                .getMethod("checkStatus").invoke(speechRecognizer);
            Futures.addCallback(future, new FutureCallback<Integer>() {
                @Override
                public void onSuccess(Integer status) {
                    int statusValue;
                    if (status == com.google.mlkit.genai.common.FeatureStatus.UNAVAILABLE) {
                        statusValue = 0;
                    } else if (status == com.google.mlkit.genai.common.FeatureStatus.DOWNLOADABLE) {
                        statusValue = 1;
                    } else if (status == com.google.mlkit.genai.common.FeatureStatus.DOWNLOADING) {
                        statusValue = 2;
                    } else if (status == com.google.mlkit.genai.common.FeatureStatus.AVAILABLE) {
                        statusValue = 3;
                    } else {
                        statusValue = 0;
                    }
                    result.success(statusValue);
                }

                @Override
                public void onFailure(Throwable e) {
                    result.error("SpeechRecognizerError", e.toString(), null);
                }
            }, executor);
        } catch (Exception e) {
            result.error("SpeechRecognizerError", "Failed to check status: " + e.toString(), null);
        }
    }

    private void startRecognition(MethodCall call, MethodChannel.Result result) {
        // Speech Recognition uses streaming - would need EventChannel
        result.notImplemented();
    }

    private void stopRecognition(MethodCall call, MethodChannel.Result result) {
        result.notImplemented();
    }

    private void closeSpeechRecognizer(MethodCall call) {
        String id = call.argument("id");
        Object speechRecognizer = instances.get(id);
        if (speechRecognizer == null) return;
        try {
            speechRecognizer.getClass().getMethod("close").invoke(speechRecognizer);
        } catch (Exception e) {
            // If close() doesn't exist, just remove from instances
        }
        instances.remove(id);
    }
}

File: packages/google_mlkit_genai_speech_recognition/android/src/main/kotlin/com/google_mlkit_genai_speech_recognition/SpeechRecognizer.kt
package com.google_mlkit_genai_speech_recognition

import android.content.Context
import com.google.common.util.concurrent.FutureCallback
import com.google.common.util.concurrent.Futures
import com.google.common.util.concurrent.ListenableFuture
import com.google.mlkit.genai.common.FeatureStatus
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.Executors

class SpeechRecognizer(
    private val context: Context,
) : MethodChannel.MethodCallHandler {
    private val instances = mutableMapOf<String, Any>()
    private val executor = Executors.newSingleThreadExecutor()

    companion object {
        private const val CHECK_STATUS = "genai#checkStatus"
        private const val START_RECOGNITION = "genai#startRecognition"
        private const val STOP_RECOGNITION = "genai#stopRecognition"
        private const val CLOSE = "genai#closeSpeechRecognizer"
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            CHECK_STATUS -> {
                checkStatus(call, result)
            }

            START_RECOGNITION -> {
                result.notImplemented()
            }

            STOP_RECOGNITION -> {
                result.notImplemented()
            }

            CLOSE -> {
                closeSpeechRecognizer(call)
                result.success(null)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initialize(call: MethodCall): Any =
        runCatching {
            val optionsBuilder =
                Class
                    .forName("com.google.mlkit.genai.speechrecognition.SpeechRecognizerOptions")
                    .getMethod("builder", Context::class.java)
                    .invoke(null, context)!!
            val options = optionsBuilder.javaClass.getMethod("build").invoke(optionsBuilder)!!
            Class
                .forName("com.google.mlkit.genai.speechrecognition.SpeechRecognition")
                .getMethod("getClient", options.javaClass)
                .invoke(null, options)!!
        }.getOrDefault(Any())

    private fun checkStatus(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val id = call.argument<String>("id") ?: return
        val speechRecognizer = instances.getOrPut(id) { initialize(call) }

        runCatching {
            @Suppress("UNCHECKED_CAST")
            val future =
                speechRecognizer.javaClass
                    .getMethod("checkStatus")
                    .invoke(speechRecognizer) as ListenableFuture<Int>

            Futures.addCallback(
                future,
                object : FutureCallback<Int> {
                    override fun onSuccess(status: Int?) {
                        val statusValue =
                            when (status) {
                                FeatureStatus.UNAVAILABLE -> 0
                                FeatureStatus.DOWNLOADABLE -> 1
                                FeatureStatus.DOWNLOADING -> 2
                                FeatureStatus.AVAILABLE -> 3
                                else -> 0
                            }
                        result.success(statusValue)
                    }

                    override fun onFailure(e: Throwable) {
                        result.error("SpeechRecognizerError", e.toString(), null)
                    }
                },
                executor,
            )
        }.onFailure { e ->
            result.error("SpeechRecognizerError", "Failed to check status: $e", null)
        }
    }

    private fun closeSpeechRecognizer(call: MethodCall) {
        val id = call.argument<String>("id") ?: return
        val speechRecognizer = instances.remove(id) ?: return
        runCatching { speechRecognizer.javaClass.getMethod("close").invoke(speechRecognizer) }
    }
}
